lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1252,17 +1261,22 @@ def _process_logging_relation_changed(self, relation: Relation):
         Args:
             relation: the `Relation` instance to update.
         """
+        self._update_relation_data(relation)
+        self._update_alert_rules(relation)
+        self._check_alert_rules()
+
+    def _update_relation_data(self, relation):
         if self._charm.unit.is_leader():
             relation.data[self._charm.app].update(self._promtail_binary_url)
             logger.debug("Saved promtail binary url: %s", self._promtail_binary_url)
             relation.data[self._charm.app]["endpoints"] = json.dumps(self._endpoints())
             logger.debug("Saved endpoints in relation data")
```

Comment:
I also changed this is my patch to verify things first, but it should be done here also. It's amazingly chatty for no reason whatsoever, and ends up with [weird guards like this one](https://github.com/canonical/loki-k8s-operator/blob/cc07f57faa1a15c0ad62d87ae3edb4df109aecf0/lib/charms/loki_k8s/v0/loki_push_api.py#L1505).

Imagine a consumer joins.

A  `relation-changed` event is fired from that side, and the provider responds with more updates, which triggers another change on the other side, and so on. Endlessly. We are relying on Juju inferring that the data is the same and it should _not_ fire an event to break this loop, and it's unnecessary.

Neither the Provider nor the Consumer listens to `_joined` at all, where "initial data" (such as juju topology/"metadata" from the consumer and the promtail URL on the provider) from the consumer side should be set, since we are reasonably sure that they will never change.

This is questionable at best and dangerous at worst, since the only reason this is working at is because `RelationChangedEvent` always fires on joined, and because Juju is checking for diffs in the state data to prevent a loop.

The pattern violates the basic assumptions relations [operate under](https://juju.is/docs/sdk/relations). In particular:

> Callback methods bound to this event should be the only ones that rely on remote relation settings. They should not error if the settings are incomplete, since it can be guaranteed that when the remote unit or application changes its settings, the event will fire again.

Nothing in this handler on either side needs this. For lifecycle events, sure. For `RelationChangedEvent`, this should just be in Joined.

Reply:
Can we take such issues in future refactors. The purpose of this PR is to implement the *minimal* set of changes that brings a measure of stability to the integration tests. The whole point of doing this is so that we can progress the development of the Loki charm. As mentioned in the PR description various github issue tickets have been opened to consider further improvements to Loki operator.

Reply:
Not to put too fine a point on it, but we **do not know the "minimal" set of changes**. Catching exceptions in the integration tests so they turn green does not address what the "minimal" set of changes may be, and we cannot stand on that as an example unless the exception handling is _removed_ and they still pass.

Reply:
I have explained in PR description why adding the exception handlers is not so that the tests "turn green". It is to allow the test and all subsequent tests in the same module to run to completion and check all their asserts in the process. It is those asserts that decide if the tests should be green or not.

Once again please do create additional issues for any other improvements you feel necessary. This purpose of this PR is not to fix all the problems of Juju, python-libjuju and pytest-operator.

Reply:
This is middle-ground position which does not make sense.

To "fix" the integration tests:

* Replace `cassandra` with `loki-tester`
* Fix `_endpoints`
* Get `loki-tester` built without the error-ed alert rules nested in the path
* Either `faulty-loki-tester` or `cos-config` (with the re-created `error_alerts`) branch.
* Use an older version of `python-libjuju` and `pytest-operator` which doesn't throw `asyncio` exceptions all the way up the stack.

That's pretty much it. The rest of the changes around common event handling/etc are nice to have, but they are not a baseline requirement if that is our goal.

As repeatedly stated, if the "scale to `0/1` problem" happens _only_ with charms using `loki_push_api`, is basic differential diagnosis that this is not necessarily a problem with Juju. That is the "Installing agent..." bug we see all over the place.

Trapping exceptions around failing to tear down the model is like waiting for your kettle to boil while the house burns down. Allowing the tests in the same module to run to completion isn't meaningful considering that one of the intended test results was "destroy these applications". Otherwise, we could just remove the relations, deploy brand-new charms, and move on with the "leftover" charms related to nothing and alive in the model.

The fact that we very explicitly `remove()` and then wait is an assertion in and of itself, in that it will elevate problems in `relation-departed|broken` when those applications _fail_ to be removed and the state of the model never matches what we expect.

It is, very purely,"turning the tests green", especially if Juju 2.9.29 resolves it. Then we should just use that and cut the exception handling.

Reply:
- Have you tried all the above suggested changes ? 

- Do they produce integration tests that run to completion and at least a shade more robust than we currently have ?

- Have you identified what about the Loki operator causes the "scale to `0/1`" issue as you call it ? I see it is juju instability issues mentioned in the PR description. I see no evidence to support that this is specific to Loki. 

- Without such concrete results aren't we chasing ghosts ?